lighthouse-cli/test/fixtures/static-server.js

```diff
@@ -21,136 +21,177 @@ const HEADER_SAFELIST = new Set(['x-robots-tag', 'link']);
-      'Content-Security-Policy': `default-src 'none';`,
-    });
-    sendResponse(200, html);
-    return;
+class Server {
+  constructor() {
+    this._server = http.createServer(this._requestHandler.bind(this));
+    /** @type {(data: string) => string} */
+    this._dataTransformer = null;
   }
 
-  if (filePath.startsWith('/dist/viewer')) {
-    // Rewrite lighthouse-viewer paths to point to that location.
-    absoluteFilePath = path.join(__dirname, '/../../../', filePath);
+  getRawServer() {
+    return this._server;
   }
 
-  // Disallow file requests outside of LH folder
-  const filePathDir = path.parse(absoluteFilePath).dir;
-  if (!filePathDir.startsWith(lhRootDirPath)) {
-    return readFileCallback(new Error('Disallowed path'));
+  /**
+   * @param {=number} port
+   * @param {=string} hostname
+   */
+  listen(port, hostname) {
+    this._server.listen(port, hostname);
   }
 
-  // Check if the file exists, then read it and serve it.
-  fs.exists(absoluteFilePath, fsExistsCallback);
-
-  function fsExistsCallback(fileExists) {
-    if (!fileExists) {
-      return sendResponse(404, `404 - File not found. ${filePath}`);
-    }
-    fs.readFile(absoluteFilePath, 'binary', readFileCallback);
+  /**
+   * @param {(err?: Error) => void} cb
+   */
+  close(cb) {
+    this._server.close(cb);
   }
 
-  function readFileCallback(err, file) {
-    if (err) {
-      console.error(`Unable to read local file ${absoluteFilePath}:`, err);
-      return sendResponse(500, '500 - Internal Server Error');
-    }
-    sendResponse(200, file);
+  /**
+   * @param {(data: string) => string} fn
+   */
+  setDataTransformer(fn) {
+    this._dataTransformer = fn;
   }
 
-  function sendResponse(statusCode, data) {
-    const headers = {'Access-Control-Allow-Origin': '*'};
-
-    const contentType = mime.lookup(filePath);
-    const charset = mime.lookup(contentType);
-    // `mime.contentType` appends the correct charset too.
-    // Note: it seems to miss just one case, svg. Doesn't matter much, we'll just allow
-    // svgs to fallback to binary encoding. `Content-Type: image/svg+xml` is sufficient for our use case.
-    // see https://github.com/jshttp/mime-types/issues/66
-    if (contentType) headers['Content-Type'] = mime.contentType(contentType);
-
-    let delay = 0;
-    let useGzip = false;
-    if (queryString) {
-      const params = new URLSearchParams(queryString);
-      // set document status-code
-      if (params.has('status_code')) {
-        statusCode = parseInt(params.get('status_code'), 10);
-      }
+  _requestHandler(request, response) {
+    const requestUrl = parseURL(request.url);
+    const filePath = requestUrl.pathname;
+    const queryString = requestUrl.search && parseQueryString(requestUrl.search.slice(1));
+    let absoluteFilePath = path.join(__dirname, filePath);
+
+    const sendResponse = (statusCode, data) => {
+      // Used by Smokerider.
+      if (this._dataTransformer) data = this._dataTransformer(data);
```

Comment:
can we get a single usage of this somewhere in the repo so we don't accidentally break it by not understanding its purpose? :)

Reply:
do you just mean a test? I don't have a use case for core :)

Reply:
yeah sure :)

Reply:
anything that would fail travis if I broke this works for me 😆

Reply:
done!